mono/metadata/w32file-unix.c

```diff
@@ -1874,16 +1874,15 @@ static gboolean share_allows_open (struct stat *statbuf, guint32 sharemode,
 			return(FALSE);
 		}
 
-		if (((file_existing_access & GENERIC_READ) &&
-		     !(sharemode & FILE_SHARE_READ)) ||
-		    ((file_existing_access & GENERIC_WRITE) &&
-		     !(sharemode & FILE_SHARE_WRITE))) {
-			/* New share mode doesn't match up */
-			mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_IO_LAYER_FILE, "%s: Access mode prevents open: requested share: 0x%" PRIx32 ", file has access: 0x%" PRIx32, __func__, sharemode, file_existing_access);
+		if (file_existing_share != sharemode &&
+				((sharemode | FILE_SHARE_READ) != sharemode) &&
```

Comment:
Why do we need this?  It's checking that the new requested sharing mode `sharemode` doesn't include `FileShare.Read`.  Why?

Isn't it enough to just check that `file_existing_share != sharemode` (new is different from existing) and `(file_existing_share | sharemode) == file_existing_share` (new it at most as lax as existing - it doesn't include).  That together means that new `sharemode` is a strict subset of the existing `file_existing_share`.  And therefore we take the branch and return FALSE.

Reply:
This change/code confuses me.
Why do the share modes have to line up, as long as they grant each other access?
Don't sharemodes get anded upon successful open, to weigh against later opens?
I might try it myself if that not too rude.

like:
  - open readwrite, share readwrite
  - open read, share read; success, resulting share is read 
  - open readwrite; fail
 
or:
    - open read, share read
    -  open read, share readwrite; success, resulting share is read
    - open readwrite; fail

Note this cannot reall workout underlying kernel support.
We end up presumably only enforcing it within the process, amongst mono users, which is a lot less than the original intent.

Reply:
@lambdageek From our conversation, the `((sharemode | FILE_SHARE_READ) != sharemode) &&` line is to close the gap w/ ReadWrite.  Without it, FIle.Copy would fail (ReadWrite to Read). 

The more I look at this and the corefx version, the more I think we should just follow in kind. 
 Which is, FileShare.None locks exclusively and that's about it.

https://github.com/dotnet/corefx/blob/b990ece3f8cd7999ad62483f1bd814ec586895de/src/Common/src/CoreLib/System/IO/FileStream.Unix.cs#L76-L88

Reply:
Yeah, since it requires kernel/filesystem to really support, i.e. to play "correctly" with non-mono callers.
A coworker of mine years ago suggested this sharing stuff, while it seems like a good idea, isn't really all that great even on Windows and perhaps is pointless. Unix might have its flaws, but this missing feature you don't hear about much.

My simple attempt broke tests.
https://github.com/mono/mono/pull/16271

Reply:
The check is pretty simple now.

Reply:
Need to "subscribe" to that file/line and get a notification if they decide to work on it.
Maybe the AI will notice the copied code and notify us, or update ours.